### PR TITLE
Audit logging to log request body, response body or headers for some …

### PR DESCRIPTION
**Pull request by yaojiefeng**
…REST endpoints

Build: http://builds.cask.co/browse/CDAP-DUT5276-3

### CODE REVIEW

File: cdap-common/src/main/java/co/cask/cdap/common/security/AuditEnum.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.common.security;
+
+/**
+ * Declares which content needs to be audit logged.
+ */
+public enum AuditEnum {
```
Comment (anew) [inline cdap-common/src/main/java/co/cask/cdap/common/security/AuditEnum.java]: Better not call this AuditEnum. We don't generally end the name of an enum type in Enum. Maybe AuditDetail, or if you know something better.

File: cdap-common/src/main/java/co/cask/cdap/common/security/AuditEnum.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  HEADER,
```
Comment (anew) [inline cdap-common/src/main/java/co/cask/cdap/common/security/AuditEnum.java]: should this be HEADERS?

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterAuditLookUp.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+      return HttpMethod.POST;
+    }
+    if (method.isAnnotationPresent(DELETE.class)) {
+      return HttpMethod.DELETE;
+    }
+    return null;
+  }
+
+  private List<Class<?>> getAllHandlerClasses() throws IOException {
+    ClassLoader cl = getClass().getClassLoader();
```
Comment (anew) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterAuditLookUp.java:146]: this assumes that all handlers are in the class path. For SDK, that is probably true. Have to tested it in a cluster?

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/SecurityAuthenticationHttpHandler.java [hunk 4]
```diff
@@ -171,6 +181,25 @@ private boolean validateSecuredInterception(ChannelHandlerContext ctx, HttpReque
...
+          Map<String, String> headers = new HashMap<>();
+          for (String headerName : headerNames) {
+            headers.put(headerName, msg.getHeader(headerName));
+          }
+          logEntry.setHeaders(headers);
+        }
+        if (auditLogContent.isLogRequestBody()) {
+          ChannelBuffer body = msg.getContent();
+          if (body.readable()) {
+            logEntry.setRequestBody(body.toString());
```
Comment (anew) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/SecurityAuthenticationHttpHandler.java]: what happens if not? Does that mean there is no body?

**Review by anew — COMMENTED** [review]
changes look mostly good. Please see my few comments.

Do you have insights into the performance impact?

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/SecurityAuthenticationHttpHandler.java (hunk 4 above)
Comment (yaojiefeng) [reply cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/SecurityAuthenticationHttpHandler.java]: Yes. That means readerIndex >= writerIndex, which means nothing we can read from the ChannelBuffer

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterAuditLookUp.java (hunk 3 above)
Comment (yaojiefeng) [reply cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterAuditLookUp.java:146]: Will test first on sdk, and then on a cluster

**Comment by yaojiefeng** [comment]
In this PR, I have not separated the audit logging from the `SecurityAuthenticationHttpHandler`, since we need the user id for the audit log. Will think about a design for it and do this in a separate PR.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java [hunk 5]
```diff
@@ -445,9 +446,11 @@ private BodyConsumer deployApplication(final HttpResponder responder,
       protected void onFinish(HttpResponder responder, File uploadedFile) {
         try {
           // deploy app
-          applicationLifecycleService.deployAppAndArtifact(namespace, appId, artifactId, uploadedFile,
-                                                           configString, createProgramTerminator());
-          responder.sendString(HttpResponseStatus.OK, "Deploy Complete");
+          ApplicationWithPrograms app =
+            applicationLifecycleService.deployAppAndArtifact(namespace, appId, artifactId, uploadedFile, configString,
+                                                             createProgramTerminator());
+          responder.sendString(HttpResponseStatus.OK, String.format("Successfully deployed app %s",
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java:452]: What happens if the deploy fails? Does the response contain any information about the app name?

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/SecurityAuthenticationHttpHandler.java [hunk 6]
```diff
@@ -367,6 +379,19 @@ private AuditLogEntry getLogEntry(ChannelHandlerContext ctx) {
...
 
+  private String findResponseBody(ChannelBuffer channelBuffer, boolean isFirstChunk) {
+    if (!isFirstChunk) {
+      return channelBuffer.readable() ? channelBuffer.toString(Charsets.UTF_8) : "";
+    }
+    int bufferEnd = channelBuffer.writerIndex();
+    int start = channelBuffer.indexOf(channelBuffer.readerIndex(), bufferEnd, TWO_SEPERATOR_FINDER);
+
+    if (start >= 0) {
+      return channelBuffer.toString(start + 4, bufferEnd - start - 4, Charsets.UTF_8).trim();
```
Comment (anew) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/SecurityAuthenticationHttpHandler.java:387]: what is the 4?

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/SecurityAuthenticationHttpHandler.java [hunk 7]
```diff
@@ -384,4 +409,22 @@ public boolean find(ChannelBuffer buffer, int guessedIndex) {
       return true;
     }
   };
+
+  private static final ChannelBufferIndexFinder TWO_SEPERATOR_FINDER = new ChannelBufferIndexFinder() {
+    private byte[] lineSeperator  = "\r\n\r\n".getBytes(Charsets.UTF_8);
```
Comment (anew) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/SecurityAuthenticationHttpHandler.java:411]: is it guaranteed that there is always \r? Or is it possible that there is simply "\n\n"?

**Comment by rohitsinha54** [comment]
Also, add some unit test for this.

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterAuditLookUp.java (hunk 3 above)
Comment (anew) [reply cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterAuditLookUp.java:146]: How do you validate that on a cluster the classpath contains all the handlers? If it doesn't there will not be an error. But for some paths the audit log will be incomplete.

**Comment by anew** [comment]
I made another pass, only a few questions left.

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/SecurityAuthenticationHttpHandler.java (hunk 6 above)
Comment (yaojiefeng) [reply cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/SecurityAuthenticationHttpHandler.java:387]: It is the length of the two line seperators '\r\n\r\n'. We do not want them to be included in the response body.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 5 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java:452]: No. It only has the message from the exceptions

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/SecurityAuthenticationHttpHandler.java (hunk 7 above)
Comment (yaojiefeng) [reply cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/SecurityAuthenticationHttpHandler.java:411]: From https://tools.ietf.org/html/rfc7230#section-2.7.3, the body is after two CRLFs, which means two `\r\n`.

**Comment by yaojiefeng** [comment]
@anew I do not know why I cannot reply on that comment.. For cluster, I think we need to have a integration test to verify that we get all the REST endpoints for audit logging.

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterAuditLookUp.java (hunk 3 above)
Comment (yaojiefeng) [reply cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterAuditLookUp.java:146]: For cluster, I think we need to have a integration test to verify that we get all the REST endpoints for audit logging.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java (hunk 5 above)
Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java:452]: ok

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/SecurityAuthenticationHttpHandler.java (hunk 6 above)
Comment (anew) [reply cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/SecurityAuthenticationHttpHandler.java:387]: a comment would be nice. The name TWO_SEPARATOR_FINDER does not indicate that it finds four characters :)

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/SecurityAuthenticationHttpHandler.java (hunk 7 above)
Comment (anew) [reply cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/SecurityAuthenticationHttpHandler.java:411]: ok

**Comment by anew** [comment]
Can you add a debug log along these lines: "Audit log lookup: bootstrapped with <n> paths"? At least that allows us to manually verify the number is the same. Also, please open a follow-up Jira for an integration test.

**Comment by yaojiefeng** [comment]
JIRA created for integration test: https://issues.cask.co/browse/CDAP-8123

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterAuditLookUp.java [hunk 8]
```diff
@@ -120,9 +121,12 @@ private void createMatcher() {
                                                               auditContents.contains(AuditDetail.REQUEST_BODY),
                                                               auditContents.contains(AuditDetail.RESPONSE_BODY),
                                                               headerNames);
+        LOG.debug("Audit log lookup: bootstrapped with path: {}", completePath);
```
Comment (anew) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterAuditLookUp.java]: make this trace? I think this is too verbose even at debug level.

**Comment by anew** [comment]
Is there also a way to unit-test the audit log itself? Perhaps configure a logback appender that writes to a temp file, then read the temp file?

**Comment by anew** [comment]
@yaojiefeng please see my remaining two comments.

**Comment by yaojiefeng** [comment]
Will try to have another unit test in a separate PR since I am not sure I can get it done before I leave.

**Comment by anew** [comment]
LGTM

**Pull request merged**